service/src/main/java/com/epam/esm/services/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,230 @@
+package com.epam.esm.services.impl;
+
+import com.epam.esm.exceptions.DataNotFoundException;
+import com.epam.esm.exceptions.WrongParameterException;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.services.GiftCertificateService;
+import com.epam.esm.services.TagService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.mapping.PropertyReferenceException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.exceptions.ExceptionCodesConstants.NOT_FOUND_GIFT_CERTIFICATE;
+import static com.epam.esm.exceptions.ExceptionCodesConstants.WRONG_PARAMETER;
+
+/**
+ * Implementation of the {@link GiftCertificateService} interface.
+ * Provides operations related to GiftCertificates, such as creation, update, deletion, and retrieval.
+ */
+@Service
+@Slf4j
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    private final GiftCertificateRepository repository;
+    private final TagService tagService;
+
+    public GiftCertificateServiceImpl(GiftCertificateRepository repository, TagService tagService) {
+        this.repository = repository;
+        this.tagService = tagService;
+    }
+
+    /**
+     * Save new instants of gift certificate and save tags (if they are new) and creating their pairs
+     *
+     * @param giftCertificate - new Entity
+     * @return {@link GiftCertificate} with generated id
+     */
+    @Transactional
+    @Override
+    public GiftCertificate saveGiftCertificate(GiftCertificate giftCertificate) throws WrongParameterException {
```

Comment:
It would be better if you used entities only in service layer. DTOs should be used in Controller layer.